### PR TITLE
New build and deployment workflow

### PR DESCRIPTION
**Pull request by severinsimmler**
This PR introduces a few new tools and workflows, but makes building and deploying wheels super easy:

- Packaging and dependency management with [Poetry](https://python-poetry.org/) via `pyproject.toml` and `poetry.lock`
- Continuous integration and deployment with GitHub Actions
- Building wheels for different Python versions and platforms with [cibuildwheel](https://cibuildwheel.readthedocs.io/en/stable/)

@ariddell, you'd have to add your PyPI username and password as secrets to the repository (or is this already available?) that we can upload the wheels to PyPI with Poetry. If you are not familiar with Poetry, they have [pretty good documentation](https://python-poetry.org/docs/cli/#publish).

You can also try out building the wheels locally:

```
$ poetry install --only build && poetry run cibuildwheel --platform linux --output-dir dist
```

This will produce wheels for all Python versions specified in the `pyproject.toml`:

```
$ tree dist/
dist/
├── lda-2.1.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
├── lda-2.1.0-cp39-cp39-musllinux_1_1_x86_64.whl
├── lda-2.1.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
├── lda-2.1.0-cp310-cp310-musllinux_1_1_x86_64.whl
├── lda-2.1.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
├── lda-2.1.0-cp311-cp311-musllinux_1_1_x86_64.whl
├── lda-2.1.0-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
└── lda-2.1.0-cp312-cp312-musllinux_1_1_x86_64.whl
```

This is just the first PR, there are a few more things I'd like to add :) For example:

- Type hints
- Refactor the unit tests and use `pytest`
- Including the example corpora only optional (as this bloats the package a bit)
- ...

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+[tool.ruff]
+line-length = 99
+
+[tool.ruff.per-file-ignores]
+"__init__.py" = ["F401"]
+
+[tool.cibuildwheel]
+build = "cp39-* cp310-* cp311-* cp312-*"
+skip = ["pp*", "*i686", "*win32"]
```
Comment (severinsimmler) [inline pyproject.toml:46]: We skip these because there are currently some issues, see: https://github.com/numpy/numpy/issues/24703#issuecomment-1731269424

**Comment by severinsimmler** [comment]
Sorry, I think I have confused @ariddell with @riddella 😬

**Comment by riddella** [comment]
This looks really promising! Thank you for working on this.

Would you mind breaking this up into three or more distinct pull requests? There are a few significant changes that I think would be better considered independently.

- The switch to poetry
- All the test changes (e.g., replacing oslotest with unittest)
- The github actions / cibuildwheel changes

p.s. Don't worry about the confusion. It's the same person. (I need to transfer this repository from @ariddell (my graduate school account) to @riddella (my current account).)

**Comment by severinsimmler** [comment]
Cool, thanks! And makes absolutely sense to split this into three PRs.

**Comment by severinsimmler** [comment]
Closing this in favor of https://github.com/lda-project/lda/pull/123 and all subsequent PRs to come

**Pull request closed without merging**